### PR TITLE
Fix AudioLibrary.GetSongs missing songs when using limits

### PR DESCRIPTION
**Pull request by DaveTBlake**
Problems have been reported in recent builds with AudioLibrary.GetSongs when using limits to return songs in sections. The total number of songs returned not matching the number displayed in Kodi.

The problem was in `CMusicDatabase::GetSongsFullByWhere` introduced by #8306 to return accurate artist data for the songs by querying a join of songview and songartistview. This was not correctly applying the limit criteria (to just songs not the join), and needed to be a left join to cover those songs that do not have artists.

Thanks @Millencolin007 for doing the leg work and pinning this issue down so well for me to fix.

I think this is safe enough to be back ported to Jarvis, but would like some testing first obviously. It is a music library issue, but only shows up when using JSON to get songs from a full library.

### CODE REVIEW

File: xbmc/music/MusicDatabase.cpp [hunk 1]
```diff
@@ -4038,8 +4038,10 @@ bool CMusicDatabase::GetSongsFullByWhere(const std::string &baseDir, const Filte
     int total = -1;
 
     std::string strSQL = "SELECT %s FROM songview ";
-    if (artistData) // Get data from song and song_artist tables to fully populate songs with artists
-      strSQL = "SELECT %s FROM songview JOIN songartistview on songartistview.idsong = songview.idsong ";
+    if (artistData)
+      // Get data from song and song_artist tables to fully populate songs with artists. 
+      //Some songs may not have artists so Lfet join
```
Comment (Razzeee) [inline xbmc/music/MusicDatabase.cpp]: typo in left

Comment (DaveTBlake) [inline xbmc/music/MusicDatabase.cpp]: Yes it can really happen.
Bug was found by someome with exactly that in their database, real life data. Elsewhere JM used left joins, and he knew what he was doing. My oversight here.

**Comment by xhaggi** [comment]
looks sane

**Comment by DaveTBlake** [comment]
Testing feedback from @Millencolin007 shows this working well. 

Once typo fixed are we OK to backport?  Would like Jarvis to return all the songs to the remotes, not miss the odd one! @MartijnKaijser ?

**Comment by MartijnKaijser** [comment]
if others are ok

**Comment by xhaggi** [comment]
sure, it's non-critical change and fix an issue.

File: xbmc/music/MusicDatabase.cpp [hunk 2]
```diff
@@ -4104,7 +4113,7 @@ bool CMusicDatabase::GetSongsFullByWhere(const std::string &baseDir, const Filte
     {
       unsigned int targetRow = (unsigned int)it->at(FieldRow).asInteger();
       const dbiplus::sql_record* const record = data.at(targetRow);
-
```
Comment (Montellese) [inline xbmc/music/MusicDatabase.cpp]: Please keep all these cosmetic changes in separate commits as they have nothing to do with the actual fix.

**Comment by Razzeee** [comment]
+1 for this.
Thank you!

**Comment by MilhouseVH** [comment]
This doesn't work with MySQL. The following JSON query

```
2016-02-02 18:20:24.127909:MainThread: libStats.JSON SOCKET REQUEST: [{"jsonrpc": "2.0", "params": {"limits": {"start": 0, "end": 1}}, "method": "AudioLibrary.GetSongs", "id": "libStats"}]
2016-02-02 18:20:24.139020:MainThread: libStats.BUFFER RECEIVED (len 85)
2016-02-02 18:20:24.139077:MainThread: libStats.PARSING JSON DATA: {"error":{"code":-32603,"message":"Internal error."},"id":"libStats","jsonrpc":"2.0"}
```

now produces the error:
```
18:18:50 65401.003906 T:1694233504   ERROR: SQL: [MyMusic59] Undefined MySQL error: Code (1235)
                                            Query: SELECT songview.*, songartistview.*  FROM songview LEFT JOIN songartistview on songartistview.idsong = songview.idsong  WHERE songview.idsong in (SELECT idsong FROM songview  LIMIT 1)
18:18:50 65401.007812 T:1694233504   ERROR: GetSongsFullByWhere() failed
```

which is...
```
Error Code: 1235. This version of MySQL doesn't yet support 'LIMIT & IN/ALL/ANY/SOME subquery'
```

This is with a MySQL 5.5 server.

I'm not entirely sure what this query is trying to achieve, but using a subquery will work in MySQL, and also works in SQLite (this isn't exactly ideal, but I can't find a better solution):

```
SELECT songview.*, songartistview.*
FROM songview LEFT JOIN songartistview on songartistview.idsong = songview.idsong
WHERE songview.idsong IN (SELECT idsong FROM (SELECT idsong FROM songview LIMIT 1) as t)
```

The following patch will fix the problem:
```
diff --git a/xbmc/music/MusicDatabase.cpp b/xbmc/music/MusicDatabase.cpp
index 7c08865..15f2334 100644
--- a/xbmc/music/MusicDatabase.cpp
+++ b/xbmc/music/MusicDatabase.cpp
@@ -4079,7 +4079,7 @@ bool CMusicDatabase::GetSongsFullByWhere(const std::string &baseDir, const Filte
         strSQL += strSQLExtra;
       else
         //Apply where clause and limits to songview, then join as mutiple records in result set per song
-        strSQL += " WHERE songview.idsong in (SELECT idsong FROM songview " + strSQLExtra + ")";
+        strSQL += " WHERE songview.idsong IN (SELECT idsong FROM (SELECT idsong FROM songview " + strSQLExtra + ") as t)";
     }
 
     CLog::Log(LOGDEBUG, "%s query = %s", __FUNCTION__, strSQL.c_str());
```

Patch: http://sprunge.us/KbFY
